horovod/run/run.py

```diff
@@ -0,0 +1,405 @@
+    args_list = [[ssh_command_format.format(host=host_address,
+                                            ssh_port_arg=ssh_port_arg)]
+                 for host_address in host_addresses]
+    ssh_exit_codes = \
+        threads.execute_function_multithreaded(exec_command,
+                                               args_list)
+    ssh_successful_to_all_hosts = True
+    for index, ssh_status in six.iteritems(ssh_exit_codes):
+        exit_code, stderr, stdout = ssh_status[0], ssh_status[1], ssh_status[2]
+        if exit_code != 0:
+            print("ssh not successful for host {host}.".format(
+                host=host_addresses[index]))
+            print("stderr:\n{stderr}\nstdout:\n{stdout}".format(stdout=stdout,
+                                                                stderr=stderr))
+            ssh_successful_to_all_hosts = False
+    if not ssh_successful_to_all_hosts:
+        exit(1)
+    return True
+
+
+def _launch_task_servers(host_addresses, driver_addresses, num_hosts, tmout,
+                         key, ssh_port=None):
+    """
+    executes the task server and service client task for registration on the
+    hosts.
+    :param host_addresses: list of addresses. for example,
+        ['worker-0','worker-1']
+        ['10.11.11.11', '10.11.11.12']
+    :type host_addresses: list(string)
+    :param driver_addresses: map of interfaces and their address and port for
+    the service. For example:
+        {
+            'lo': [('127.0.0.1', 34588)],
+            'docker0': [('172.122.10.1', 34588)],
+            'eth0': [('11.111.33.73', 34588)]
+        }
+    :type driver_addresses: map
+    :param num_hosts:
+    :type num_hosts: int
+    :param tmout:
+    :type tmout: horovod.spark.util.timeout.Timeout
+    :return:
+    :rtype:
+    """
+
+    def _exec_command(command):
+        stdout_w = six.StringIO()
+        stderr_w = six.StringIO()
+        try:
+            exit_code = safe_shell_exec.execute(command,
+                                                stdout=stdout_w,
+                                                stderr=stderr_w)
+            if exit_code != 0:
+                print(
+                    "Launching horovodrun task function was not successful.")
+                print("stderr from host:\n {stderr}".format(
+                    stderr=stderr_w.getvalue()))
+                print("stdout from host:\n {stdout}".format(
+                    stdout=stdout_w.getvalue()))
+                os._exit(exit_code)
+        finally:
+            stdout_w.close()
+            stderr_w.close()
+        return exit_code
+
+    if ssh_port:
+        ssh_port_arg = "-p {ssh_port}".format(ssh_port=ssh_port)
+    else:
+        ssh_port_arg = ""
+
+    command_format = \
+        'ssh -o StrictHostKeyChecking=no {host} {ssh_port_arg} \'source ~/.bashrc; {python} -m horovod.run.horovod_task_fn {index} {driver_addresses} {num_hosts} {timeout} {key}\''
```

Comment:
Why do we need `source ~/.bashrc`?  Some people don't use bash

Reply:
removed.